### ISSUE TITLE
Observers for Removed events - Should fire before removal

### ISSUE THREAD

**Issue by Fredosixx** [SketchUp/api-issue-tracker#402]
All observers detecting a removal of entity, instances, definitions, etc... fired AFTER the removal takes place.

As a consequence, the object or id passed in argument of the observer event method is referring to a deleted entity and is TOTALLY USELESS, since we cannot query anything about it.

**Request**: have observer events firing BEFORE removal so that we can do meaningful things on the objects or entities. Having the 'after' event is just a plus, but as said, totally useless.

I think anyway **the whole Observer API should be completely revisited**. The way it has been designed is very often unusable, and in many cases buggy, even if some bugs have been fixed along the SketchUp versions. At a minimum, any event-driven framework triggers explicit methods _before_ and _after_ the event happens. Also, there is no clarity today on what you can do and not do within an observer method. The concept of Observer is very powerful, but the current implementation is flawed, to say the least.

**Comment by DanRathbun** [comment]
> Request: have observer events firing BEFORE removal so that we can do meaningful things on the objects or entities. Having the 'after' event is just a plus, but as said, totally useless.

I agree with you and we've made this argument in the past, and got a canned response along the lines of "it's normal in any code environment for event handlers fire **after** the event occurs".

They cannot be changed now without breaking extensions "in the wild".

The only alternative is to file API requests for specific `onBefore...` or `onPre...` observer callbacks.
We've done requests in the past and some have been implemented, others not.

> I think anyway the whole Observer API should be completely revisited. 

This, pragmatically speaking, is never going to get to the top of the list. Especially since they are now wanting to shorten release cycles.

So again, file individual requests for specific "before" callbacks along with use cases, and you'll greatly increase chances of implementation.

**Comment by Fredosixx** [comment]
As indicated, my request is about **a new callback method firing BEFORE the event**, for
  - EntityObserver
  - EntitiesObserver
  - DefinitionObserver
  - DefinitionsObserver
and more generally for any of these observer methods atht deal with deletion and pass today argument which are no longer valid and usable.

And in the documentation, indicate the does and donts that can used within an observer method.

I agree that a complete or even partial overhaul of Observers is out of sight...!

**Comment by DanRathbun** [comment]
Okay, (I'm not against the need) but still, this is not separate filed issues per desired callback method.

**Comment by kengey** [comment]
I would like to see both `pre`and `post` events, for kind of every observer there is.

**Comment by thomthom** [comment]
Can you all describe in more detail the scenarios of when you want pre-events? Providing some more context to what you are trying to achieve.

Also - what would the purpose of post-events be? Seems to be the normal event and post-even is pretty much the same...

**Comment by Fredosixx** [comment]
The typical scenario is to update internal context in Ruby before deletion, for instance
  - the position and other information about the object to be deleted
  - the attributes of the object deleted when applicable

As said, **otherwise the observer event is pretty useless**, as we cannot know at all what is deleted from the second argument passed in the method.

This should apply to all '...removed' observer events
  - Definitions
  - Definitions
  - Entities
  - Entity
  - Materials
  - Pages

The current event is a POST event. Changing its name would simply make it clearer, as you did for the Save Model event.

**Comment by DanRathbun** [comment]
> The current event is a POST event. Changing its name would simply make it clearer, as you did for the Save Model event.

In effect the old callback method names get aliased, but the core then needs to poll the observer objects for 2 "post" event callbacks instead of 1. (Otherwise [renaming instead of aliasing methods] would break a lot of code "in the wild".)

This is what was done for the `ModelObserver` save callbacks.

**Comment by Fredosixx** [comment]
Renaming to 'Post...' is really a detail, versus documenting properly the API, in particular for Observers, but not just that....

And backward compatibility is a very respectable mantra I adhere to. But many SU versions have screwed up things (like SU2017), not by changing method names.

**Comment by thomthom** [comment]
Fredo - why is it that you need to read data from entities being deleted? Can you elaborate a bit more on the overall workflow?

**Comment by kengey** [comment]
> Fredo - why is it that you need to read data from entities being deleted? Can you elaborate a bit more on the overall workflow?

To clean up internal state, related to the deleted entity.
for example. However, I can imagine that we bind data based on the entity_id, and that is given upon the event.

**Comment by Fredosixx** [comment]
It is relatively frequent that extensions maintain a framework with meta data about the model, either with attributes and/or Ruby methods.

So knowing which entity is deleted is paramount for keeping integrity of the two representations.

In FredoGuides, I wanted to be informed of guides being deleted, get their position, so that I could re-establish them after deletion. Instead, I have to periodically check the position of guides, since the event of their deletion is totally uninformative.

Now, to answer your question by another one, **how useful is a method telling you 'something has been deleted' with telling you what?**

**Comment by thomthom** [comment]
> In FredoGuides, I wanted to be informed of guides being deleted, get their position, so that I could re-establish them after deletion. Instead, I have to periodically check the position of guides, since the event of their deletion is totally uninformative.

So the key here is to identify what what deleted. Accessing the properties is a way to workaround not being able to ID it?

`Sketchup::EntitiesObserver#onElementRemoved` returns the `entityID` of what was erased. (If I recall this did originally return the entity in very early SketchUp versions.)
Would providing the ID for the "removed/erased" events  be a solution to this?
http://ruby.sketchup.com/Sketchup/EntitiesObserver.html#onElementRemoved-instance_method****

**Comment by Fredosixx** [comment]
> So the key here is to identify what what deleted. Accessing the properties is a way to workaround not being able to ID it?

Not exactly and rather the reverse. We want to access the properties of the deleted object.

When the observer event fires, the object is **_already deleted_**. So having its entityID or its ruby object reference does not help, since all call to Sketchup API methods will generate an error "reference to a deleted entity".

**Comment by thomthom** [comment]
> Although it would make more sense now that we have persistent ids, to give that id.

Except that PID isn't supported on all entities.

**Comment by DanRathbun** [comment]
Generally, I support having new `onBefore...` /  `onPre...` callbacks, but not the `onAfter...` / `onPost...` aliasing of post event callbacks as this just makes the engine do double duty after the event.

**Comment by kengey** [comment]
> Generally, I support having new `onBefore...` / `onPre...` callbacks, but not the `onAfter...` / `onPost...` aliasing of post event callbacks as this just makes the engine do double duty after the event.

yes. pre and post, or pre and on, how you want to name it.

**Comment by kengey** [comment]
> > Although it would make more sense now that we have persistent ids, to give that id.
> 
> Except that PID isn't supported on all entities.

yes, I know. During the period PID were introduced I was not that active one SketchUp developments. Before PIDs we used our own ID mechanism, afterwards of course we were happy they finally got there. Weird to see they are not present on all sorts of entities, that would make sense though.

Regarding the need to access an entity's attribute dictionary: I do that a lot, but it is also an antipattern. Better is to use the model's dictionary solely and treat that as you would treat a sql database. You point your extra information to the entity instead of making it part of the entity. This, every extra info attribute you have, has a key that points to the entity. But then again, that would only work if ids were persistent...

**Comment by Fredosixx** [comment]
So, as said in the [first posts](https://github.com/SketchUp/api-issue-tracker/issues/402#issuecomment-564121556), the request boils down to **have a onBeforeRemoved event** (or whatever event method name).
